srv/src/main/java/my/bookshop/handlers/ReviewServiceHandler.java

```diff
@@ -0,0 +1,61 @@
+package my.bookshop.handlers;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+import com.sap.cds.Row;
+import com.sap.cds.ql.Select;
+import com.sap.cds.services.ErrorStatuses;
+import com.sap.cds.services.ServiceException;
+import com.sap.cds.services.cds.CdsService;
+import com.sap.cds.services.handler.EventHandler;
+import com.sap.cds.services.handler.annotations.Before;
+import com.sap.cds.services.handler.annotations.ServiceName;
+
+import cds.gen.catalogservice.CatalogService_;
+import cds.gen.my.bookshop.Reviews;
+import cds.gen.my.bookshop.Reviews_;
+import cds.gen.reviewservice.ReviewService_;
+import my.bookshop.MessageKeys;
+
+@Component
+@ServiceName(ReviewService_.CDS_NAME)
+public class ReviewServiceHandler implements EventHandler {
+
+	private CdsService catalogService;
+
+	ReviewServiceHandler(@Qualifier(CatalogService_.CDS_NAME) CdsService catalogService) {
+		this.catalogService = catalogService;
+	}
+
+	@Before(event = { CdsService.EVENT_CREATE, CdsService.EVENT_UPDATE })
+	public void beforeAddReview(Stream<Reviews> reviews) {
```

Comment:
Do we need this handler if the entity is readonly?

Reply:
In the ReviewService the `Reviews` entity is not readonly, only within the CatalogService.

Reply:
Yes, right. But why don't we use the `ReviewService` then in the `CatalogServiceHandler` to add the review?

Reply:
Well, yeah, the `ReviewService` implementation came after I've implemented the `CatalogServiceHandler` and the `AddReview` action in there. The separate `ReviewService` was thought of as an experiment to try more customization options in the Fiori Elements UI, as they seem to be quite restrictive for bound actions. And in the end I left it, as it demonstrates some other things quite nicely.

And in the `CatalogService` the `PersistenceService` was already injected, so I just reused it.

But if you prefer I can inject the `ReviewService` and use that one instead. That way I can remove the Review validation logic in there, as it's already handled by the `ReviewService`.

Reply:
Yeah, that's what I also thought. We wouldn't duplicate the validation logic. Also later on, we could maybe even separate the two services in separate micro-services if we want to showcase that.